File: lintcode/python/0872_kill_process.py
#!/usr/bin/python -t

# BFS

from collections import deque

class Solution:
    """
    @param pid: the process id
    @param ppid: the parent process id
    @param kill: a PID you want to kill
    @return: a list of PIDs of processes that will be killed in the end
    """
    def killProcess(self, pid, ppid, kill):
        # Write your code here
        
        mapping = {}
        
        for i, parent in enumerate(ppid):
            if parent not in mapping:
                mapping[parent] = []
            mapping[parent].append(pid[i])
            
        q = deque([kill])
        ret = []
        
        while len(q) > 0:
            p = q.popleft()
            #print p, mapping
            ret.append(p)
            #print ret, type(p)
            if p in mapping:
                q += mapping[p]
                
        return ret
        

if __name__ == '__main__':
    s = [1,3,10,5]
    v = [3,0,5,3]
    k = 5
    ss = Solution()
    print "answer is\n"
    print ss.killProcess(s, v, k)

File: lintcode/golang/0872_kill_process.go
/**
 * @param pid: the process id
 * @param ppid: the parent process id
 * @param kill: a PID you want to kill
 * @return: a list of PIDs of processes that will be killed in the end
 */
 
//import "fmt"
import "sort"

func killProcess (pid []int, ppid []int, kill int) []int {
    // Write your code here
    ppidToPid := make(map[int][]int)
    
    for i, pp := range ppid {
        //p, _ := ppidToPid[pp]
        //if !ok {
        //    p = make([]int, 0)
        //}
        ppidToPid[pp] = append(ppidToPid[pp], pid[i])
    }
    
    //fmt.Println(ppidToPid)
    
    ret := []int{}
    
    q := make([]int, 0)
    q = append(q, kill)
    
    for len(q) != 0 {
        //fmt.Println(q)
        p := q[0]
        q = q[1:len(q)]
        ret = append(ret, p)
        for _, pd := range ppidToPid[p] {
            q = append(q, pd)
        }
    }
    
    sort.IntSlice(ret).Sort()
    return ret
}
